src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_contract_functions.cpp

```diff
@@ -0,0 +1,454 @@
+    pure_contract_symbol.name, spec_assigns_function_id, empty_typet());
+
+  const auto &spec_function_id = spec_function_symbol.name;
+
+  auto &spec_code_type = to_code_type(spec_function_symbol.type);
+
+  exprt::operandst lambda_parameters;
+
+  if(code_with_contract.return_type().id() != ID_empty)
+  {
+    // use a dummy symbol for __CPROVER_return_value
+    // which does occur in the assigns clause anyway
+    lambda_parameters.push_back(
+      symbol_exprt("dummy_return_value", code_with_contract.return_type()));
+  }
+
+  for(const auto &param_id : spec_code_type.parameter_identifiers())
+  {
+    lambda_parameters.push_back(ns.lookup(param_id).symbol_expr());
+  }
+
+  // fetch the goto_function to add instructions to
+  goto_functiont &goto_function =
+    goto_model.goto_functions.function_map.at(spec_function_id);
+  goto_programt &body = goto_function.body;
+
+  for(const auto &assigns_expr : code_with_contract.assigns())
+  {
+    auto expr = to_lambda_expr(assigns_expr).application(lambda_parameters);
+    expr.add_source_location() = assigns_expr.source_location();
+    if(can_cast_expr<conditional_target_group_exprt>(expr))
+    {
+      encode_assignable_target_group(
+        to_conditional_target_group_expr(expr), body);
+    }
+    else
+    {
+      encode_assignable_target(expr, body);
+    }
+  }
+
+  body.add(goto_programt::make_end_function(spec_function_symbol.location));
+
+  goto_model.goto_functions.update();
+
+  inline_and_check_warnings(spec_function_id);
+
+  PRECONDITION_WITH_DIAGNOSTICS(
+    utils.has_no_loops(spec_function_id),
+    "loops in assigns clause specification functions must be unwound before "
+    "model instrumentation");
+
+  goto_model.goto_functions.update();
+}
+
+void dfcc_dsl_contract_functionst::encode_assignable_target_group(
+  const conditional_target_group_exprt &group,
+  goto_programt &dest)
+{
+  const source_locationt &source_location = group.source_location();
+
+  // clean up side effects from the condition expression if needed
+  cleanert cleaner(goto_model.symbol_table, log.get_message_handler());
+  exprt condition(group.condition());
+  if(has_subexpr(condition, ID_side_effect))
+    cleaner.clean(condition, dest, language_mode);
+
+  // Jump target if condition is false
+  auto goto_instruction = dest.add(
+    goto_programt::make_incomplete_goto(not_exprt{condition}, source_location));
+
+  for(const auto &target : group.targets())
+    encode_assignable_target(target, dest);
+
+  auto label_instruction = dest.add(goto_programt::make_skip(source_location));
+  goto_instruction->complete_goto(label_instruction);
+}
+
+void dfcc_dsl_contract_functionst::encode_assignable_target(
+  const exprt &target,
+  goto_programt &dest)
+{
+  const source_locationt &source_location = target.source_location();
+
+  if(can_cast_expr<side_effect_expr_function_callt>(target))
+  {
+    // A function call target
+    //
+    // `foo(params)`
+    //
+    // becomes
+    //
```

Comment:
Nit pick: those blank lines in the comment make it harder to fit many lines of context on one screen. The entire comment actually made it harder for me to understand what was going on on this occasion.

Reply:
opinions may differ but it think it actually can actually help my future self navigate the code and get a global understanding as opposed to the very noisy code that builds the expression which quite frankly requires deep inspection to figure out. 
 
```c
    const auto &funcall = to_side_effect_expr_function_call(target);
    code_function_callt code_function_call(to_symbol_expr(funcall.function()));
    auto &arguments = code_function_call.arguments();
    for(auto &arg : funcall.arguments())
      arguments.emplace_back(arg);
    dest.add(
      goto_programt::make_function_call(code_function_call, source_location));

```

Reply:
compressed it to one line